File: ios/Tests/BitchatMeshTests/MeshListenerTests.swift
import XCTest
import CoreBluetooth
@testable import BitchatMesh

final class MeshListenerTests: XCTestCase {
    private final class RecordingListener: MeshListener {
        var messages: [BitchatMessage] = []
        var received: [BitchatMessage] = []
        var peerListUpdates: [[PeerID]] = []
        var found: [PeerID] = []
        var lost: [PeerID] = []
        var connected: [PeerID] = []
        var disconnected: [PeerID] = []
        var established: [PeerID] = []
        var rssi: [(PeerID, Int)] = []
        var started = false
        var stopped = false
        var sent: [(String?, String?)] = []
        var deliveryAcks: [(String, String)] = []
        var readReceipts: [(String, String)] = []
        var verifyChallenges: [PeerID] = []
        var verifyResponses: [PeerID] = []
        var bluetoothStates: [CBManagerState] = []
        var publicMessages: [(PeerID, String, String)] = []
        var noisePayloads: [(PeerID, NoisePayloadType)] = []
        var transferEvents: [TransferProgressEvent] = []
        var filesReceived: [(PeerID, String)] = []
        var incomingExpectation: XCTestExpectation?
        var progressExpectation: XCTestExpectation?

        func onMessageReceived(_ message: BitchatMessage) {
            messages.append(message)
        }

        func onReceived(_ message: BitchatMessage) {
            received.append(message)
        }

        func onSent(messageID: String?, recipientPeerID: String?) {
            sent.append((messageID, recipientPeerID))
        }

        func onPeerListUpdated(_ peers: [PeerID]) {
            peerListUpdates.append(peers)
        }

        func onFound(_ peerID: PeerID) {
            found.append(peerID)
        }

        func onLost(_ peerID: PeerID) {
            lost.append(peerID)
        }

        func onPeerConnected(_ peerID: PeerID) {
            connected.append(peerID)
        }

        func onPeerDisconnected(_ peerID: PeerID) {
            disconnected.append(peerID)
        }

        func onEstablished(_ peerID: PeerID) {
            established.append(peerID)
        }

        func onRSSIUpdated(peerID: PeerID, rssi: Int) {
            self.rssi.append((peerID, rssi))
        }

        func onStarted() {
            started = true
        }

        func onStopped() {
            stopped = true
        }

        func onDeliveryAck(messageID: String, recipientNickname: String, timestamp: Date) {
            deliveryAcks.append((messageID, recipientNickname))
        }

        func onReadReceipt(messageID: String, recipientNickname: String, timestamp: Date) {
            readReceipts.append((messageID, recipientNickname))
        }

        func onVerifyChallenge(peerID: PeerID, payload: Data, timestamp: Date) {
            verifyChallenges.append(peerID)
        }

        func onVerifyResponse(peerID: PeerID, payload: Data, timestamp: Date) {
            verifyResponses.append(peerID)
        }

        func onBluetoothStateUpdated(_ state: CBManagerState) {
            bluetoothStates.append(state)
        }

        func onPublicMessageReceived(from peerID: PeerID, nickname: String, content: String, timestamp: Date, messageID: String?) {
            publicMessages.append((peerID, nickname, content))
        }

        func onNoisePayloadReceived(from peerID: PeerID, type: NoisePayloadType, payload: Data, timestamp: Date) {
            noisePayloads.append((peerID, type))
        }

        func onTransferProgress(transferId: String, sent: Int, total: Int, completed: Bool) {
            transferEvents.append(TransferProgressEvent(transferId: transferId, sent: sent, total: total, completed: completed))
            progressExpectation?.fulfill()
        }

        func onFileReceived(peerID: PeerID, fileName: String, fileSize: Int, mimeType: String, localURL: URL) {
            filesReceived.append((peerID, fileName))
            incomingExpectation?.fulfill()
        }
    }

    func testMeshListenerEndpoints() {
        let mesh = MeshManager()
        let listener = RecordingListener()
        mesh.listener = listener

        let peerA = PeerID(str: "peerA")
        let peerB = PeerID(str: "peerB")
        let message = BitchatMessage(sender: "alice", content: "hi", timestamp: Date(), isRelay: false)
        mesh.didReceiveMessage(message)
        XCTAssertEqual(listener.messages.count, 1)
        XCTAssertEqual(listener.received.count, 1)

        mesh.sendPrivateMessage("secret", to: peerA, recipientNickname: "bob", messageID: "msg-1")
        XCTAssertEqual(listener.sent.map { $0.0 }, ["msg-1"])
        mesh.didUpdatePeerList([peerA, peerB])
        XCTAssertEqual(listener.peerListUpdates.count, 1)

        mesh.didConnectToPeer(peerA)
        mesh.didDisconnectFromPeer(peerA)
        XCTAssertEqual(listener.found, [peerA])
        XCTAssertEqual(listener.connected, [peerA])
        XCTAssertEqual(listener.lost, [peerA])
        XCTAssertEqual(listener.disconnected, [peerA])

        let delivered = DeliveryStatus.delivered(to: "bob", at: Date())
        let read = DeliveryStatus.read(by: "bob", at: Date())
        mesh.didUpdateMessageDeliveryStatus("msg-1", status: delivered)
        mesh.didUpdateMessageDeliveryStatus("msg-2", status: read)
        XCTAssertEqual(listener.deliveryAcks.map { $0.0 }, ["msg-1"])
        XCTAssertEqual(listener.readReceipts.map { $0.0 }, ["msg-2"])

        mesh.didReceiveNoisePayload(from: peerA, type: .verifyChallenge, payload: Data([0x01]), timestamp: Date())
        mesh.didReceiveNoisePayload(from: peerA, type: .verifyResponse, payload: Data([0x02]), timestamp: Date())
        XCTAssertEqual(listener.verifyChallenges, [peerA])
        XCTAssertEqual(listener.verifyResponses, [peerA])
        XCTAssertEqual(listener.noisePayloads.map { $0.1 }, [.verifyChallenge, .verifyResponse])

        mesh.didUpdateBluetoothState(.poweredOn)
        XCTAssertEqual(listener.bluetoothStates, [.poweredOn])

        mesh.didReceivePublicMessage(from: peerB, nickname: "bob", content: "hello", timestamp: Date(), messageID: "m1")
        XCTAssertEqual(listener.publicMessages.count, 1)

        listener.incomingExpectation = expectation(description: "incoming file")
        mesh.didReceiveFileTransfer(peerID: peerA, fileName: "file.txt", fileSize: 12, mimeType: "text/plain", localURL: URL(fileURLWithPath: "/tmp/file.txt"))
        wait(for: [listener.incomingExpectation!], timeout: 1.0)
        XCTAssertEqual(listener.filesReceived.count, 1)
    }

    func testTransferProgressForwarding() {
        let mesh = MeshManager()
        let listener = RecordingListener()
        mesh.listener = listener

        listener.progressExpectation = expectation(description: "progress")
        listener.progressExpectation?.expectedFulfillmentCount = 3

        TransferProgressManager.shared.start(id: "t1", totalFragments: 2)
        TransferProgressManager.shared.recordFragmentSent(id: "t1")
        TransferProgressManager.shared.recordFragmentSent(id: "t1")

        wait(for: [listener.progressExpectation!], timeout: 2.0)
        XCTAssertEqual(listener.transferEvents.count, 3)
    }

    func testStartStopCallbacks() {
        let mesh = MeshManager()
        let listener = RecordingListener()
        mesh.listener = listener

        mesh.start(nickname: "tester")
        mesh.stop()

        XCTAssertTrue(listener.started)
        XCTAssertTrue(listener.stopped)
    }
}

File: android/bitchat-mesh/src/test/kotlin/com/bitchat/mesh/MeshListenerTests.kt
package com.bitchat.mesh

import android.content.Context
import com.bitchat.android.mesh.BluetoothMeshDelegate
import com.bitchat.android.mesh.TransferProgressEvent
import com.bitchat.android.mesh.TransferProgressManager
import com.bitchat.android.model.BitchatMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.UnconfinedTestDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.setMain
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.util.Date
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@OptIn(ExperimentalCoroutinesApi::class)
class MeshListenerTests {
    private val dispatcher = UnconfinedTestDispatcher()

    @BeforeEach
    fun setUp() {
        Dispatchers.setMain(dispatcher)
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun delegateForwardsMeshEvents() {
        val meshManager = MeshManager(mockContext())
        val listener = RecordingListener()
        meshManager.setListener(listener)

        val delegate = buildDelegate(meshManager)
        val message = BitchatMessage(
            sender = "alice",
            content = "hi",
            timestamp = Date(),
            senderPeerID = "peerA"
        )

        delegate.didReceiveMessage(message)
        delegate.didEstablishSession("peerA")
        delegate.didUpdateRSSI("peerA", -42)
        delegate.didReceiveDeliveryAck("msg-1", "peerA")
        delegate.didReceiveReadReceipt("msg-2", "peerA")
        delegate.didReceiveVerifyChallenge("peerA", byteArrayOf(0x01), 123L)
        delegate.didReceiveVerifyResponse("peerA", byteArrayOf(0x02), 124L)
        delegate.didReceiveFileTransfer("peerA", "file.txt", 10L, "text/plain", "/tmp/file.txt")

        assertEquals(1, listener.receivedMessages.size)
        assertEquals(listOf("peerA"), listener.established)
        assertEquals(listOf("peerA" to -42), listener.rssiUpdates)
        assertEquals(listOf("msg-1" to "peerA"), listener.deliveryAcks)
        assertEquals(listOf("msg-2" to "peerA"), listener.readReceipts)
        assertEquals(listOf("peerA"), listener.verifyChallenges)
        assertEquals(listOf("peerA"), listener.verifyResponses)
        assertEquals(listOf("peerA" to "file.txt"), listener.filesReceived)
    }

    @Test
    fun delegateForwardsPeerListChanges() {
        val meshManager = MeshManager(mockContext())
        val listener = RecordingListener()
        meshManager.setListener(listener)
        val delegate = buildDelegate(meshManager)

        delegate.didUpdatePeerList(listOf("peerA", "peerB"))
        delegate.didUpdatePeerList(listOf("peerB", "peerC"))

        assertEquals(listOf("peerA", "peerB", "peerC"), listener.found)
        assertEquals(listOf("peerA", "peerB", "peerC"), listener.connected)
        assertEquals(listOf("peerA"), listener.lost)
        assertEquals(listOf("peerA"), listener.disconnected)
        assertEquals(2, listener.peerListUpdates)
    }

    @Test
    fun sendCallsInvokeOnSent() {
        val meshManager = MeshManager(mockContext())
        val listener = RecordingListener()
        meshManager.setListener(listener)

        meshManager.sendBroadcastMessage("hi")
        meshManager.sendPrivateMessage("secret", "peerA", "alice", "msg-123")

        assertEquals(2, listener.sent.size)
        assertEquals(null to null, listener.sent.first())
        assertEquals("msg-123" to "peerA", listener.sent.last())
    }

    @Test
    fun startStopEndpoints() {
        val listener = RecordingListener()

        listener.onStarted()
        listener.onStopped()

        assertTrue(listener.started)
        assertTrue(listener.stopped)
    }

    @Test
    fun transferProgressIsForwarded() = runBlocking {
        val meshManager = MeshManager(mockContext())
        val listener = RecordingListener()
        meshManager.setListener(listener)

        TransferProgressManager.start("transfer-1", 2)
        TransferProgressManager.progress("transfer-1", 1, 2)
        TransferProgressManager.complete("transfer-1", 2)

        assertTrue(listener.transferLatch.await(2, TimeUnit.SECONDS))
        val ordered = listener.transferEvents.sortedBy { it.sent }
        assertEquals(
            listOf(
                TransferProgressEvent("transfer-1", 0, 2, false),
                TransferProgressEvent("transfer-1", 1, 2, false),
                TransferProgressEvent("transfer-1", 2, 2, true)
            ),
            ordered
        )
    }

    private fun buildDelegate(meshManager: MeshManager): BluetoothMeshDelegate {
        val method = MeshManager::class.java.getDeclaredMethod("buildDelegate")
        method.isAccessible = true
        return method.invoke(meshManager) as BluetoothMeshDelegate
    }

    private fun mockContext(): Context {
        val context = mock<Context>()
        whenever(context.applicationContext).thenReturn(context)
        return context
    }

    private class RecordingListener : MeshListener {
        val receivedMessages: MutableList<BitchatMessage> = mutableListOf()
        val established: MutableList<String> = mutableListOf()
        val rssiUpdates: MutableList<Pair<String, Int>> = mutableListOf()
        val deliveryAcks: MutableList<Pair<String, String>> = mutableListOf()
        val readReceipts: MutableList<Pair<String, String>> = mutableListOf()
        val verifyChallenges: MutableList<String> = mutableListOf()
        val verifyResponses: MutableList<String> = mutableListOf()
        val filesReceived: MutableList<Pair<String, String>> = mutableListOf()
        val found: MutableList<String> = mutableListOf()
        val lost: MutableList<String> = mutableListOf()
        val connected: MutableList<String> = mutableListOf()
        val disconnected: MutableList<String> = mutableListOf()
        var peerListUpdates: Int = 0
        val sent: MutableList<Pair<String?, String?>> = mutableListOf()
        var started = false
        var stopped = false
        val transferEvents: MutableList<TransferProgressEvent> = mutableListOf()
        val transferLatch = CountDownLatch(3)

        override fun onMessageReceived(message: BitchatMessage) {
            receivedMessages.add(message)
        }

        override fun onPeerListUpdated(peers: List<String>) {
            peerListUpdates += 1
        }

        override fun onFound(peerID: String) {
            found.add(peerID)
        }

        override fun onLost(peerID: String) {
            lost.add(peerID)
        }

        override fun onConnected(peerID: String) {
            connected.add(peerID)
        }

        override fun onDisconnected(peerID: String) {
            disconnected.add(peerID)
        }

        override fun onEstablished(peerID: String) {
            established.add(peerID)
        }

        override fun onRSSIUpdated(peerID: String, rssi: Int) {
            rssiUpdates.add(peerID to rssi)
        }

        override fun onStarted() {
            started = true
        }

        override fun onStopped() {
            stopped = true
        }

        override fun onSent(messageID: String?, recipientPeerID: String?) {
            sent.add(messageID to recipientPeerID)
        }

        override fun onDeliveryAck(messageID: String, recipientPeerID: String) {
            deliveryAcks.add(messageID to recipientPeerID)
        }

        override fun onReadReceipt(messageID: String, recipientPeerID: String) {
            readReceipts.add(messageID to recipientPeerID)
        }

        override fun onVerifyChallenge(peerID: String, payload: ByteArray, timestampMs: Long) {
            verifyChallenges.add(peerID)
        }

        override fun onVerifyResponse(peerID: String, payload: ByteArray, timestampMs: Long) {
            verifyResponses.add(peerID)
        }

        override fun onTransferProgress(transferId: String, sent: Int, total: Int, completed: Boolean) {
            transferEvents.add(TransferProgressEvent(transferId, sent, total, completed))
            transferLatch.countDown()
        }

        override fun onFileReceived(peerID: String, fileName: String, fileSize: Long, mimeType: String, localPath: String) {
            filesReceived.add(peerID to fileName)
        }
    }
}
